qubes/qmemman/systemstate.py

```diff
@@ -18,210 +17,211 @@
 #
 # You should have received a copy of the GNU General Public
 # License along with this library; if not, see <https://www.gnu.org/licenses/>.
+
 import functools
 import logging
 import os
 import time
-
-import xen.lowlevel
+import xen.lowlevel  # pylint: disable=import-error
 from pathlib import Path
+from typing import Optional
 
 import qubes.qmemman
 from qubes.qmemman.domainstate import DomainState
 
-
-no_progress_msg = "VM refused to give back requested memory"
-slow_memset_react_msg = "VM didn't give back all requested memory"
+BALLOON_DELAY = 0.1
+XEN_FREE_MEM_LEFT = 50 * 1024 * 1024
+XEN_FREE_MEM_MIN = 25 * 1024 * 1024
+MEM_OVERHEAD_FACTOR = 1.0 / 1.00781
+CHECK_PERIOD_S = 3
+CHECK_MB_S = 100
+MIN_TOTAL_MEMORY_TRANSFER = 150 * 1024 * 1024
+MIN_MEM_CHANGE_WHEN_UNDER_PREF = 15 * 1024 * 1024
 
 
-class SystemState(object):
-    def __init__(self):
+class SystemState:
+    def __init__(self) -> None:
         self.log = logging.getLogger("qmemman.systemstate")
         self.log.debug("SystemState()")
 
-        self.domdict = {}
-        self.xc = None
-        self.xs = None
+        self.dom_dict: dict[str, DomainState] = {}
+        self.xc: xen.lowlevel.xc.xc = None
+        self.xs: xen.lowlevel.xs.xs = None
+        self.all_phys_mem: int = 0
 
-    def init(self):
+    def init(self) -> None:
         self.xc = xen.lowlevel.xc.xc()
         self.xs = xen.lowlevel.xs.xs()
-        self.BALOON_DELAY = 0.1
-        self.XEN_FREE_MEM_LEFT = 50 * 1024 * 1024
-        self.XEN_FREE_MEM_MIN = 25 * 1024 * 1024
         # Overhead of per-page Xen structures, taken from OpenStack
         # nova/virt/xenapi/driver.py
-        # see https://wiki.openstack.org/wiki/XenServer/Overhead
-        # we divide total and free physical memory by this to get
-        # "assignable" memory
-        self.MEM_OVERHEAD_FACTOR = 1.0 / 1.00781
+        # See https://wiki.openstack.org/wiki/XenServer/Overhead
+        # We divide total and free physical memory by this to get "assignable"
+        # memory
         try:
-            self.ALL_PHYS_MEM = int(
-                self.xc.physinfo()["total_memory"]
-                * 1024
-                * self.MEM_OVERHEAD_FACTOR
+            self.all_phys_mem = int(
+                self.xc.physinfo()["total_memory"] * 1024 * MEM_OVERHEAD_FACTOR
             )
         except xen.lowlevel.xc.Error:
-            self.ALL_PHYS_MEM = 0
+            pass
+
+    def xs_wrapper(self, key, value=None) -> Optional[str]:
+        if value:
+            self.xs.write("", key, value)
+            return None
+        return self.xs.read("", key)
```

Comment:
Please don't, this doesn't save much (just one empty parameter), but makes it harder to see if it's a read or write (and easier to make a mistake).
OTOH the get_xs_path helper is useful change.